### PR TITLE
Experimental: allow parent Block to consume child Block's toolbar

### PR DESCRIPTION
**Pull request by getdave**
Several recent use cases have suggested that it would be valuable to a parent Block rendering child blocks into `InnerBlocks` to be able to ~~"consume"~~ capture the toolbars of the children (in reverse: the children have their toolbars "proxied" to the parent toolbar area to be displayed).

The key benefits of this are:

1. the toolbar remains fixed to the top left avoiding the "toolbar overload" problem associated with having toolbars "follow" the cursor around as the child blocks are selected.
2. the toolbar is rendered in a DOM node divorced from the child Block which works around difficult `overflow: hidden` problems.
3. nested blocks are easier to use because the toolbar is always in a consistent position and not effected by the inner block layout.

~Two specific use cases for this might be:~

~1. Nav Block - see https://github.com/WordPress/gutenberg/pull/18336~
~2. Social Block~

**Update**: for use case please see [Joen's comment below](https://github.com/WordPress/gutenberg/pull/18440#issuecomment-565391080).

In general however, the pattern also lends itself well to any deeply nested Block hierarchies. For example, on the current `master` branch, try nesting a lot of Columns and see how you get on with nested toolbars. It isn't a ideal experience. 

By deferring the _rendering_ of all toolbars to the root block of the hierarchy we provide a better experience.

The PR is experimental and achieves the effect by:

* Rendering a `Slot` for ChildToolbars in all _parents_ of the _currently selected_ Block.
* Conditionally rendering the `BlockContextualToolbar` into the ChildToolbars Slot.

The reason for using Slot/Fill is due to needing to capture events triggered on the childtoolbar within the React component tree rather than the DOM tree. If we allow for the events to bubble in the DOM it triggers selection of the underlying parent Block (which is where the Slot is rendered) which breaks a lot of toolbar functionality.

This works. There are some quirks that would need to be worked out.

I'm looking for feedback about:

1. Whether this goes too far in breaking the existing UI patterns (see Table Block which does something similar _UI wise_ - although it doesn't use `InnerBlocks` or a Parent -> Child relationship)
2. What a11y concerns there might be about having the child toolbar divorced from the block itself and could these be mitigated.
3. Where my code is not considering edge cases.

## How has this been tested?
Manual testing for now.

## Testing

1. Add Nav Menu
2. Add some items
3. Add some nested items
4. Click between the Nav items and see all the child toolbars proxied to the top level Nav Menu.
5. Test keyboard shortcut for jumping to closest toolbar works (ie: `⌥ + F10`)

Also try [switching out the boolean props on `InnerBlocks`](https://github.com/WordPress/gutenberg/blob/18e6909dcbdafaa99916020215494a1388e3b184/packages/block-library/src/navigation-menu/edit.js#L243) on the Nav Menu Block and verify that the documented behaviour works as described.

Also try applying these props to the `InnerBlocks` of the Columns Block. Try nesting!

## Screenshots 
Screen Capture on 2019-11-15 at 14-52-08

## Types of changes
New feature (non-breaking change which adds functionality) 

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. .

### CODE REVIEW

**Comment by shaunandrews** [comment]
I'm a fan of this, as it keeps the tools in a consistent place and keeps things feeling a little simpler overall — less moving parts. Also avoids the awkward state where the toolbar can't move all the way and never really lines up with anything.

**Comment by karmatosed** [comment]
I really like the idea of this as it has a toolbar that is relevant to what is happening.

**Comment by retrofox** [comment]
Nice suggestion. It will affect [how to deal with mobile](https://github.com/WordPress/gutenberg/issues/18401).

**Comment by mtias** [comment]
This is discussed here, for reference: #17267

File: packages/block-library/src/navigation-menu-item/index.js
Comment (mtias) [inline packages/block-library/src/navigation-menu-item/index.js]: I feel this could be a property of `InnerBlocks` instead.

Comment (getdave) [reply packages/block-library/src/navigation-menu-item/index.js]: Yes i was thinking about this last night. I can see how it might work. I will try on Wednesday.

**Comment by talldan** [comment]
I like the idea of this, I think it could work well where a parent block has a close relationship to its inner blocks, just like the nav and social links blocks. 

Visually it might need some adjustment. I found two things, the left border seemed strange on the selected inner block now that it doesn't align with the toolbar. Secondly, I didn't think it was quite as clear that the inner block was selected. A bit of adjustment to those borders could probably solve both issues:
Screen Shot 2019-11-12 at 5 17 54 pm

It also seems like the more menu on inner blocks can't be selected. I realise the PR is probably pretty early in development, though.

From a code point of view, the terms `consumeChildToolbar` and `proxyToolbarToParent` seem like they could be simplified a bit for block implementors. Something like `pinToolbarToParentBlock` might be clearer.

**Comment by jasmussen** [comment]
Nice work thank you!

This is what I see:

parent-toolbar

_The shift in height on menu item select is a separate issue to this PR, but worth looking at_.

I think this is good:

- It scales better to a lot of contexts
- As a prop, it is likely this will be helpful to a ton of exotic blocks we have yet to see — imagine a parent block where child blocks can be rotated or skewed, it would really help if the toolbar wasn't attached to those child blocks individually
- Even though we are separately exploring ways to improve selecting nested blocks, this can further benefit specific blocks, at a block authors discretion, giving more tools to enable a good experience.

In that vein, 👍 👍

I would echo @talldan in suggesting that we might need some visual adjustment. There was something accidentally nice in having the toolbar centered over each menu item — that relationship would be nice to explore restoring. And also, the style of each selected child item can probably be holistically considered. But none of those explorations, IMO, should happen in this PR. What this does is essentially replace the hacky CSS that was written for the Navigation Block and Social Links blocks, to "mimic" this behavior in absence of this prop.

In that vein, it might be worth looking at how much of that code we can now remove and clean up, now that it's official. For example, #17877 might make the horizontal margin work simpler, and lines 14-68 in https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/social-links/editor.scss#L14 and 1-55 in https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/navigation-menu/editor.scss SHOULD be made redundant by this PR. I'm not they are, but they should be :)

File: packages/block-library/src/navigation-menu-item/index.js
Comment (getdave) [reply packages/block-library/src/navigation-menu-item/index.js]: Done!

**Comment by getdave** [comment]
> It also seems like the more menu on inner blocks can't be selected.

Agreed. Looks like hovering/clicking anywhere outside the format controls triggers the underlying parent Block to be selected.

I'm wondering whether to wait on the below to land before we try and spend too much time fixing the current implementation:

https://github.com/WordPress/gutenberg/pull/17875

File: packages/block-editor/src/components/block-list/block.js
Comment (jorgefilipecosta) [inline packages/block-editor/src/components/block-list/block.js]: This component and the comments are duplicating the one we have bellow. Could we join both conditions with an or and avoid this code duplication?

File: packages/block-editor/src/components/block-list/block.js
Comment (jorgefilipecosta) [inline packages/block-editor/src/components/block-list/block.js]: I think we can not rely on isRootOfHierarchy and need a more complex condition.
If the navigation block is nested several levels deep e.g: inside the columns. This logic will make the toolbar of the navigation menu item appear on the columns block while the toolbar of the navigation menu itself appears bellow on the block.
The toolbar should not appear on the top-level block but on the closes ancestor that used consumeChildToolbar.

File: packages/block-editor/src/components/inner-blocks/index.js
Comment (jorgefilipecosta) [inline packages/block-editor/src/components/inner-blocks/index.js]: Any thoughts about inheritance? E.g: if a block defines  __experimentalConsumeChildToolbars should all the descent blocks have this setting enable or it should only pass one level down if the child does not explicitly set the flag for its descendants?

File: packages/block-editor/src/components/block-list/block.js
Comment (jorgefilipecosta) [inline packages/block-editor/src/components/block-list/block.js]: BlockContextualToolbar renders the toolbars elements added by slot&fill but also things like the block switcher and more menu options etc.
This mechanism is working well for things added by the block that pass by the slot & fill, but when clicks happen outside on the more menu, for example, the parent block gets selected (as you pointed out on the core editor chat).
It is not exclusive of the more menu e.g: if we click on child block menu icon the parent gets selected, it is also impossible to use the block switcher (possible to test by changing the allowed blocks of navigation menu to allowedBlocks={ [ 'core/navigation-menu-item', 'core/group' ] } so the group transform is available).
This happens because the click event is happening on the dom of the parent. When a click occurs on the dom of a block, that block gets selected.

I think we have three solutions:
- Change the code that selects a block to be aware of this logic and do nothing if the toolbar being shown is one of the children. The code will need to check if the event is happening specifically inside the toolbar as if the click occurs outside the toolbar, the parent should get selected.
- Use IgnoreNestedEvents on the tollbar, so events do not propagate to the code that selects the block. I'm not sure if this has unintended consequences and causes regressions I am not aware of.
- Use a slot & fill mechanism. Blocks render a slot for the child toolbar. The child BlockContextualToolbar is rendered inside the fill of the parent if the flag is true. This solution, although following a different path, may solve the problem of multiple nesting levels as in slot & fill fills are rendered on the closest ancestor slot.

**Review by jorgefilipecosta — COMMENTED** [review]
Hi @getdave thank you for submitting this proposal 👍 This mechanism is very useful in some blocks we have. I did an initial review and shared some comments.

File: packages/block-editor/src/components/block-list/block.js
Comment (getdave) [reply packages/block-editor/src/components/block-list/block.js]: Yep. I noticed this as well. The conditional isn't water tight. I need to review it. Thanks for flagging.

File: packages/block-editor/src/components/block-list/block.js
Comment (getdave) [reply packages/block-editor/src/components/block-list/block.js]: Agreed. There's no point in doubling up. It's currently like it is for clarity during dev. But we need to consolidate and DRY it up.

File: packages/block-editor/src/components/block-list/block.js
Comment (getdave) [reply packages/block-editor/src/components/block-list/block.js]: You make a good point. I need to start thinking about Columns more and nesting.

> ...the closes ancestor that used consumeChildToolbar.

What tools are available to us to access the settings for each `InnerBlocks`? 

Also this assumes that we're going to have `consumeChildToolbar ` apply recursively. In hindsight, I'm thinking perhaps that would be best but I'm not 100% sure. There might be hidden use cases I'm overlooking where you'd need fine grain control.

File: packages/block-editor/src/components/inner-blocks/index.js
Comment (getdave) [reply packages/block-editor/src/components/inner-blocks/index.js]: See https://github.com/WordPress/gutenberg/pull/18440#discussion_r346248621

File: packages/block-editor/src/components/block-list/block.js
Comment (getdave) [reply packages/block-editor/src/components/block-list/block.js]: > This mechanism is working well for things added by the block that pass by the slot & fill, but when clicks happen outside on the more menu

> This happens because the click event is happening on the dom of the parent. When a click occurs on the dom of a block, that block gets selected.

You are right. This is what [I observed](https://github.com/WordPress/gutenberg/pull/18440#issuecomment-553396286).

> Change the code that selects a block to be aware of this logic and do nothing if the toolbar being shown is one of the children. 

I considered this. This would need to rely on particular DOM characteristic to detect whether the event occured within a toolbar. Not sure how robust it would be.

> Use IgnoreNestedEvents on the tollbar, so events do not propagate to the code that selects the block.

I considered this also. Was trying to grok `IgnoreNestedEvents` yesterday. May need some more advice on it.

> Use a slot & fill mechanism.

I'm in favour of exploring this as it seems to solve an number of the pain points. Need to get my head around how this might work.

File: packages/block-editor/src/components/block-list/block.js
Comment (getdave) [reply packages/block-editor/src/components/block-list/block.js]: This is no longer a problem.

Screen Shot 2019-11-15 at 13 15 48

File: packages/block-editor/src/components/block-list/block.js
Comment (getdave) [reply packages/block-editor/src/components/block-list/block.js]: I've gone with the Slot/Fill approach and I can confirm that the entire toolbar is now usable as the events are set to bubble within the React component tree rather than the DOM tree. This means the events never selects the parent Block.

Screen Capture on 2019-11-15 at 13-19-28

**Comment by getdave** [comment]
@karmatosed I'd be grateful for any input you felt able to offer here. Specifically:

* I've added the title of the Block to the Toolbar (see screenshot) **only** _if_ it's a captured child toolbar (ie: a child toolbar being rendered in the parent toolbar position). I find this makes it clearer which toolbar you are interacting with.
Screen Shot 2019-11-15 at 15 01 34

* Is there anything else we might like to do to denote the toolbar is part of the selected child block and not the parent toolbar.
* How does the usage "feel" on the Nav Block (where I've enabled it for now)?
* Any quirks or bugs you are experiencing?

**Comment by getdave** [comment]
@jorgefilipecosta I've heavily revised the implementation.

The situation is quite complex. Before we go into this, we need to remember is that it is not the _Block_ that defines whether it captures toolbars, but the prop on `InnerBlocks`. 

As things work _now_:

* the root level of the hierarchy which is an ancestor of the _selected block_ always renders a slot (note that doesnt mean anything is rendered into that Slot yet!).
* the descedant of an `InnerBlocks` with the `captureToolbar` prop defined will render it's toolbar into the Slot defined on the root parent.

Essentially then **_all_** toolbars propagate upwards to the root parent.

However, we have issues with recursion.

For example, if you have a `core/columns` Block that sets `captureToolbar` on its `InnerBlocks` then all the immediate child `core/column` Blocks' toolbars are captured. However, if you then nest another `core/columns` within one of those Columns then the following happens (not the pluralisation carefully!):

* the new (nested) `core/columns` does _not_ have its toolbar captured - this is because its parent `core/column` (singular!) does not set `captureToolbar` on _its_ `InnerBlocks`.
* the toolbar of the child `core/column` of the _nested_ `core/columns` **_is_** captured by the root level `core/columns` - this is because `core/columns` `edit.js` defines `captureToolbar` on `InnerBlocks`.

This is correct behaviour but it is unintuitive and provides a weird UX. What I think should happen is:

1. We rename the prop `captureDescdendantsToolbars` (plural).
2. Setting `captureDescdendantsToolbars` on a parent `InnerBlocks` causes _all_ child toolbars to be "captured".
3. We add another prop called `captureImmediateChildToolbar` (singular). If defined it only captures the toolbar of the immediate children of the `InnerBlocks` on which it is set.
4. `captureDescdendantsToolbars` overides `captureImmediateChildToolbar`.

The issue I have is how can I access the props of an `InnerBlocks` from a child? I don't know of any way to do this.

**Comment by mtias** [comment]
Thanks for exploring this! Even if we don't use it for the navigation block I believe this would be a great addition to the `InnerBlocks` API for some complex needs (like rotated items, as @jasmussen points out).

One thing that becomes clear with this flow (testing with `Cover`) is a need to select the parent from the toolbar within the area of the block type icon.

**Comment by getdave** [comment]
> Thanks for exploring this! Even if we don't use it for the navigation block I believe this would be a great addition to the InnerBlocks API for some complex needs

That's great to hear! I anticipate shipping this without adding to any Blocks initially so the stuff I've added to Nav will be removed prior to merge (if we get that far!).

> ... a need to select the parent from the toolbar within the area of the block type icon.

Do you mean the Block title that appears above the toolbar? Are you anticipating something like the screenshot below, where clicking on `Cover` would select the parent?

Screen Shot 2019-11-15 at 18 48 44

If I've got the wrong idea, please let me know.

File: packages/block-editor/src/components/block-list/block.js
Comment (getdave) [reply packages/block-editor/src/components/block-list/block.js]: I think this has now become redundant due to refactoring.

**Comment by getdave** [comment]
Just thought for all Blocks we really need to find the closest ancestor with the capture  prop defined. 

To do this we can make a new selector which gets all the Blocklist settings for an arraybof clientIds and returns them keyed by clientId.

We can then check this for the capture prop and then use this to display the Child toolbar Slot on that Block. 

This will ensure that a Block shows its toolbar on the closest ancestor with the prop defined.

Doesnt solve the issue where you want to propagate to the root but that could be an additional setting or a different prop. There isnt a one size fits all solution here.

**Comment by mtias** [comment]
> Do you mean the Block title that appears above the toolbar? Are you anticipating something like the screenshot below, where clicking on Cover would select the parent?

Yes, exactly. Though the current implementation could use some design improvements.

**Comment by karmatosed** [comment]
I am going to loop in @jasmussen because whilst see comment above, for me, there is a correlation across work going on right now around toolbar and block interactions.

I keep thinking about the experience of using this how it should move seamlessly but feel connected. Like when you walk through a room to the next. Which means you need a 'door' to get back.

**Comment by jasmussen** [comment]
Let me take a look at this as soon as I can.

**Comment by getdave** [comment]
Question: @karmatosed @jasmussen as this will be shipped as an "experimental" API should we try to handle all the Design and UX in _this_ PR, or shall we merge this one and then split out _new_ PRs to tackle the UX?

Bear in mind we won't be immediately enabling this feature on any Blocks that I'm aware of so it won't be exposed at all initially.

Just a thought, as I tend to end up with giant PRs that become difficult to manage (at least for me anyway) and I'm trying to make an effort to slim them down.

I appreciate I've introduced a Design/UX change here (the Block name above the toolbar on child blocks) but I'm happy to roll that back as I made sure it was a single commit.

That said, I'd _greatly_ value Design input here. Is there a Master Issue that is tracking all these toolbar upgrades and improvements because we could look to centralise the efforts there to increase visibility.

I'll defer to your expertise here. 

Thanks again 😄

**Comment by karmatosed** [comment]
I am ok with iteration if an experimental flag, we just did that for navigation. I would like issues opened to make sure it is iterated though.

**Comment by jasmussen** [comment]
So, reading up on this, it seems the thing that needs a little ux love stems from this, correct?

> One thing that becomes clear with this flow (testing with Cover) is a need to select the parent from the toolbar within the area of the block type icon.

Indeed, I think this serves two purposes: one is to actually make it possible to select the parent, which is important in complex nesting situations, and two is to _indicate that you're now looking at a child block's toolbar_. The latter is important because if the two toolbars are in the same place, which block's toolbar are you looking at?

I that vein, right at this moment I'm thinking something along these lines:

Screenshot 2019-11-19 at 10 26 00

That is, we sort of collapse the parent toolbar into a breadcrumb or back arrow ur up arrow or parent icon, and we offset the child toolbar ever so slightly. 

As you can no doubt tell from the sketch, though, this idea needs a little more time in the oven, and a little higher fidelity, and since I have some other things I need to get to today, I have set a reminder to mock this up tomorrow. 

However I also agree with Tammie, I think it's totally fine to merge this as is to keep the PR small, then I'll follow up with a mockup.

File: packages/block-editor/src/components/inner-blocks/README.md
Comment (jorgefilipecosta) [inline packages/block-editor/src/components/inner-blocks/README.md]: Do we need both use cases __experimentalCaptureDescendantsToolbars, and __experimentalCaptureChildToolbar? Any block you have in mind where we will use __experimentalCaptureChildToolbar but not __experimentalCaptureDescendantsToolbars?
If using both options would it make sense to use a single prop __experimentalCaptureToolbar, with possible values being 'ancestor', 'child'?

File: packages/block-editor/src/components/block-list/block.js
Comment (jorgefilipecosta) [inline packages/block-editor/src/components/block-list/block.js]: I think we should call createSlotFill( 'ChildToolbar' ); outside of the render function, and the components ChildToolbar, ChildToolbarSlot should also be called outside too.

File: packages/block-editor/src/components/block-list/block.js
Comment (jorgefilipecosta) [inline packages/block-editor/src/components/block-list/block.js]: I think we need to improve this logic when I type something on the navigation block so the tollbar disappears and then I press option + fn + f10, I see the toolbar render on the menu item and then move the parent. I guess forcing may also need to isForcingContextualToolbar inside the fill.

File: packages/block-editor/src/components/block-list/block.js
Comment (jorgefilipecosta) [inline packages/block-editor/src/components/block-list/block.js]: Should we render BlockContextualToolbar in a variable when it is going to be used to avoid duplicating the code?

**Review by jorgefilipecosta — COMMENTED** [review]
Thank you for the iterations, I left some comments but this seems to be working well 👍

**Comment by getdave** [comment]
I'm going to get back to this just as soon as I can. Likely this Friday.

File: packages/block-editor/src/store/selectors.js
Comment (getdave) [reply packages/block-editor/src/store/selectors.js]: I've added to `createSelector` wrapper. @jorgefilipecosta please could you verify I've followed the docs correctly here?

File: packages/block-editor/src/components/block-list/block.js
Comment (getdave) [reply packages/block-editor/src/components/block-list/block.js]: I've moved outside render. Probably going to extract to a separate file.

File: packages/block-editor/src/components/block-list/block.js
Comment (getdave) [reply packages/block-editor/src/components/block-list/block.js]: I've updated this so that the logic for both parent and child accounts for the `isForcingContextualToolbar`. Seems to work under testing for me now.

File: packages/block-editor/src/components/block-list/block.js
Comment (getdave) [reply packages/block-editor/src/components/block-list/block.js]: I've updated this.

File: packages/block-editor/src/components/inner-blocks/README.md
Comment (getdave) [reply packages/block-editor/src/components/inner-blocks/README.md]: At the current time, I can't actually think of any use case for the immediate child toolbar capture. Perhaps we should just standardise on `__experimentalCaptureDescendantsToolbars`?

**Comment by getdave** [comment]
@jorgefilipecosta Do you think it would be possible to write e2e tests for this feature? Not sure how feasible it is.

File: packages/block-editor/src/components/inner-blocks/README.md
Comment (getdave) [reply packages/block-editor/src/components/inner-blocks/README.md]: I've decided to normalise to a single API to capture all toolbars. The immediate child API is now removed.

File: packages/block-editor/src/components/block-list/block.js
Comment (jorgefilipecosta) [inline packages/block-editor/src/components/block-list/block.js]: Minor: In this condition instead of relying on an index comparison we may rely on the flag hasAncestorCapturingToolbars already defined.

File: packages/block-editor/src/store/selectors.js
Comment (jorgefilipecosta) [inline packages/block-editor/src/store/selectors.js]: I think normally we don't put selector parameters here.

**Review by jorgefilipecosta — COMMENTED** [review]
This PR seems to work as expected 👍  I'm only noticing a small visual issue when applying the flag to the navigation block:
Untitled 5

Should we try to address it?

**Comment by getdave** [comment]
> Should we try to address it?

@jorgefilipecosta I believe this to be a known issue unrelated to this PR. Not saying it shouldn't be addressed not sure this is the best place to do it just in case it is a larger problem. Let me see if I can track it down.

File: packages/block-editor/src/store/selectors.js
Comment (getdave) [reply packages/block-editor/src/store/selectors.js]: You're saying we shouldn't include `clientIds` in the array here?

I can't find any examples where this pattern is repeated so I think it must be correct. I was going on [this from the Rememo docs](https://github.com/aduth/rememo):

> ...a second function which returns the reference or array of references upon which the selector's derivation depends

I assume `clientIds` was the key factor here. Because the value should be recomputed both if the `clientIds` change or the `blockListSettings` change. 🤷‍♂

File: packages/block-editor/src/components/block-list/block.js
Comment (getdave) [reply packages/block-editor/src/components/block-list/block.js]: Thanks. Fixed.

File: packages/block-editor/src/store/selectors.js
Comment (getdave) [reply packages/block-editor/src/store/selectors.js]: I fixed this btw.

**Review by jorgefilipecosta — COMMENTED** [review]
Hi @getdave,
I'm sorry on the previous comment I shared the wrong screenshot.
The correct screenshot of the visual issue I'm facing is this one: 
Untitled


When the toolbar moves to the parent the black line at the side should also move otherwise there is a visual disconnect.

**Comment by getdave** [comment]
@jorgefilipecosta I've been thinking more about this. I'm not totally convinced that moving the indicator is optimal for user experience.

For example, if I were to do this how would I know that I've hovered/selected the child block? The only indicator would be the block movers which I feel is insufficient.

Perhaps I can suggest we simply tweak the position of the focus/select indicator so it is placed in a nicer way. See below.

https://d.pr/i/Oq79cr/MUF2JwVfHi

**Comment by jasmussen** [comment]
I'm excited for this prop to land, for all the creative blocks it will enable.

However in light of the detached block toolbar and more intentional block outlines being discussed in #18667, I think it's perhaps good to zoom out a little bit and think about the use case for this, because I think that purpose may have changed a little bit since the effort was started. 

**Initially** the effort was intended to primarily address complex UI in very tight nesting situations like **Buttons** or **Navigation**. The mover control, horizontal movement, thick borders, hover styles, dashed outlines were all obstructions to getting the menu done. However this is what Navigation looks like today:

context

It has a few rough edges and needs more love, but it is definitely usable. And the toolbar being shown in context of the Menu Item child blocks has proper context. Nesting dashed outlines are removed, hover outlines are removed, [movers are moving to the toolbar](#18685), and the toolbar is becoming detached so it won't be cropped when shown on the right. Given all those changes, it seems like neither Navigation nor Buttons blocks need this prop anymore.

**So what's the use case today?**

There are a number of blocks that don't yet exist which use nested blocks in such a complicated way that toolbars simply cannot be attached. For example, what if you could add editable blocks that could be rotated? 

Screenshot 2019-12-13 at 11 26 46

You'd still want to be able to edit the text, select it, mark parts of it bold or link them. But you would not want the toolbar to also be rotated along with all of the other RichText UI, so allowing the parent to be the provider of the toolbar seems like a way to enable such blocks. In this ase, the author would have to provide additional affordances to make the block feel coherent, which is why as a non-default prop it's an additional tool in the toolbelt.

Currently you can put the Cover block in a state where it might make sense the parent absorb the toolbar:

Screenshot 2019-12-13 at 11 32 05

In this case I tend to think the prop isn't the right answer, and that something like #18779 is probably a better answer. But nonetheless you can imagine situations like it, where the parent absorbing it is appropriate.

Another use case is blocks that featur multiple layers on top of each other, where it would make sense for the toolbar also to sit in a consistent space. Whether complicated slideshow blocks, or blocks that feature filter-layers that you have to select using the select tool or the breadcrumb bar — in these cases the toolbar for each layer might just add complexity.

**So to summarize**: the ability for a parent to absorb the child block feels like something that a bunch of as-yet-written blocks could leverage to great effect, so 👍 👍 on that. But we should probably not use this prop in any of the core blocks quite yet.

**Review by jorgefilipecosta — APPROVED** [review]
From the code point of view, this change looks good to me 👍 Thank you for the iterations you did @getdave. Given that these changes impact the block UI I guess we need a look from a designer.

**Comment by jasmussen** [comment]
> Given that these changes impact the block UI I guess we need a look from a designer.

Can you clarify this? It "just" attaches the block to the parent block when the prop is used, no?

The navigation block is not changed in this PR, which as noted previously it no longer needs:

Screenshot 2019-12-13 at 13 28 12

so 👍 👍 on that one.

**Comment by getdave** [comment]
Thanks to you both for all your help here!

**Comment by ellatrix** [comment]
What is this currently used for? I don't see any use in core.

**Comment by jasmussen** [comment]
> What is this currently used for? I don't see any use in core.

I wish Github didn't hide comments when there are enough of them. The explanation as a result was buried here: https://github.com/WordPress/gutenberg/pull/18440#issuecomment-565391080

The gist of it is this:

> There are a number of blocks that don't yet exist which use nested blocks in such a complicated way that toolbars simply cannot be attached. For example, what if you could add editable blocks that could be rotated?

You can create any number of crazy child blocks. Rotate them, squish them, filter them, invert them, make them so small or so big that an attached toolbar just doesn't make sense. 

I realize the context of removing the toolbar from the DOM potentially enabling the rotation of the block without also rotating the toolbar, but I believe there is still a use case for some blocks to have a toolbar attached to the parent block.

**Comment by jasmussen** [comment]
Another block idea: "scratchpad block". It's basically a canvas you paint on, and the block toolbar holds tools to draw. Each stroke of the brush is a child block that you can later move (absolute position) or rotate or scale.

**Pull request merged**
